src/main/java/com/es/phoneshop/model/services/impl/DefaultCartService.java

```diff
@@ -38,32 +48,33 @@ public Cart getCart(HttpServletRequest req) {
     public void add(HttpServletRequest req, long productId, int quantity) throws OutOfStockException {
         Cart cart = getCart(req);
 
-        Product product = ArrayListProductDao.getInstance().getProduct(productId);
+        Product product = productDao.getProduct(productId);
         List<CartItem> cartList = cart.getItemList();
         Optional<CartItem> optionalItem = cartList.stream()
                 .filter(item -> item.getProduct().equals(product))
                 .findAny();
 
+        CartItem item = null;
         if (optionalItem.isPresent()) {
-            CartItem item = optionalItem.get();
-            int totalQuantity = item.getQuantity() + quantity;
-            if (totalQuantity <= product.getStock()) {
-                item.setQuantity(totalQuantity);
-                saveList(req, cart);
-            } else {
-                throw new OutOfStockException(String
-                        .format("Item quantity [= %d] more than stock [= %d] of item!", totalQuantity, product.getStock()));
-            }
+            item = optionalItem.get();
+        } else {
+            item = new CartItem(product, 0);
         }
-        else {
-            if (quantity <= product.getStock()) {
-                cart.add(new CartItem(product, quantity));
-                saveList(req, cart);
-            }
-            else {
+
+        lock.lock();
+        int realStock = item.getProduct().getStock();
+        int totalClientQuantity = quantity + item.getQuantity();
+        if (totalClientQuantity <= realStock) {
+            item.getProduct().setStock(realStock - quantity);
+            lock.unlock();
```

Comment:
What is gonna happen if you won't enter this 'if' section and if some exception will be thrown? I suppose your lock will not be released